validator/db/kv/attester_protection.go

```diff
@@ -95,32 +117,112 @@ func (store *Store) CheckSlashableAttestation(
-// key by storing its signing root under the appropriate bucket as well
-// as its source and target epochs for future slashing protection checks.
-func (store *Store) ApplyAttestationForPubKey(
+// SaveAttestationForPubKey saves an attestation for a validator public
+// key for local validator slashing protection.
+func (store *Store) SaveAttestationForPubKey(
 	ctx context.Context, pubKey [48]byte, signingRoot [32]byte, att *ethpb.IndexedAttestation,
 ) error {
+	store.batchedAttestationsChan <- &attestationRecord{
+		pubKey:      pubKey,
+		source:      att.Data.Source.Epoch,
+		target:      att.Data.Target.Epoch,
+		signingRoot: signingRoot,
+	}
+	// Subscribe to be notified when the attestation record queued
+	// for saving to the DB is indeed saved. If an error occurred
+	// during the process of saving the attestation record, the sender
+	// will give us that error. We use a buffered channel
+	// to prevent blocking the sender from notifying us of the result.
+	responseChan := make(chan saveAttestationsResponse, 1)
+	defer close(responseChan)
+	sub := store.batchAttestationsFlushedFeed.Subscribe(responseChan)
+	defer sub.Unsubscribe()
+	res := <-responseChan
+	return res.err
+}
+
+// Meant to run as a background routine, this function checks whether:
+// (a) we have reached a max capacity of batched attestations in the Store or
+// (b) attestationBatchWriteInterval has passed
+// Based on whichever comes first, this function then proceeds
+// to flush the attestations to the DB all at once in a single boltDB
+// transaction for efficiency. Then, batched attestations slice is emptied out.
+func (store *Store) batchAttestationWrites(ctx context.Context) {
+	ticker := time.NewTicker(attestationBatchWriteInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case v := <-store.batchedAttestationsChan:
+			store.batchedAttestations = append(store.batchedAttestations, v)
+			if len(store.batchedAttestations) == attestationBatchCapacity {
+				log.WithField("numRecords", attestationBatchCapacity).Debug(
+					"Reached max capacity of batched attestation records, flushing to DB",
+				)
+				store.flushAttestationRecords(ctx)
+			}
+		case <-ticker.C:
+			if len(store.batchedAttestations) > 0 {
+				log.WithField("numRecords", len(store.batchedAttestations)).Debug(
+					"Batched attestation records write interval reached, flushing to DB",
+				)
+				store.flushAttestationRecords(ctx)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Flushes a list of batched attestations to the database
+// and resets the list of batched attestations for future writes.
+// This function notifies all subscribers for flushed attestations
+// of the result of the save operation.
+func (store *Store) flushAttestationRecords(ctx context.Context) {
+	err := store.saveAttestationRecords(ctx, store.batchedAttestations)
+	// If there was no error, we reset the batched attestations slice.
+	if err == nil {
```

Comment:
if there is an error while writing the data, the attestations are not reset from the slice. So if for any reason, the attestation record has some bad data all consequent flushes will continue failing here